python/cuspatial/cuspatial/core/binops/binops.py

```diff
@@ -0,0 +1,409 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.
+
+from abc import ABC, abstractmethod
+
+import cudf
+
+from cuspatial.core._column.geocolumn import GeoColumn
+from cuspatial.core.binops.contains import contains_properly
+from cuspatial.utils.column_utils import (
+    contains_only_linestrings,
+    contains_only_multipoints,
+    contains_only_points,
+    contains_only_polygons,
+    has_same_dimension,
+)
+
+
+class _binop(ABC):
+    @abstractmethod
+    def preprocess(self, op, lhs, rhs):
+        """Preprocess the input data for the binary operation. This method
+        should be implemented by subclasses. Preprocess and postprocess are
+        used to implement the discrete math rules of the binary operations.
+
+        Notes
+        -----
+        Should only be called once.
+
+        Parameters
+        ----------
+        op : str
+            The binary operation to perform.
+        lhs : GeoSeries
+            The left-hand-side of the GeoSeries-level binary operation.
+        rhs : GeoSeries
+            The right-hand-side of the GeoSeries-level binary operation.
+
+        Returns
+        -------
+        GeoSeries
+            The left-hand-side of the internal binary operation, may be
+            reordered.
+        GeoSeries
+            The right-hand-side of the internal binary operation, may be
+            reordered.
+        """
+        pass
+
+    @abstractmethod
+    def postprocess(self, op, point_indices, point_result):
+        """Postprocess the output data for the binary operation. This method
+        should be implemented by subclasses.
+
+        Postprocess converts the raw results of the binary operation into
+        the final result. This is where the discrete math rules are applied.
+
+        Parameters
+        ----------
+        op : str
+            The binary operation to post process. Determines for example the
+            set operation to use for computing the result.
+        processed : cudf.Series
+            The data after applying the fundamental binary operation.
+
+        Returns
+        -------
+        cudf.Series
+            The output of the post processing, True/False results for
+            the specified binary op.
+        """
+        pass
+
+    def __init__(self, op, lhs, rhs, align=True):
+        """Compute the binary operation `op` on `lhs` and `rhs`.
+
+        There are ten binary operations supported by cuspatial:
+        - `.equals`
+        - `.disjoint`
+        - `.touches`
+        - `.contains`
+        - `.contains_properly`
+        - `.covers`
+        - `.intersects`
+        - `.within`
+        - `.crosses`
+        - `.overlaps`
+
+        There are twenty-five ordering combinations of `lhs` and `rhs`:
```

Comment:
There are 36 and I don't think it's necessary to list all of them.

Reply:
I was thinking of using the list to track their implementations. You're right they certainly don't need to be enumerated in the public docstring! Thanks. I'll revisit this doc.